File: SpiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        int nn;
        int mm;
        int x = 0;
        vector<int>v;
        int total = n * m;
        
        while(total > 0)
        {
            nn = x;
            mm = x;
            while(mm != m)
            {
                v.push_back(matrix[nn][mm]);
                mm++;
                total --;
            }
            
            mm--;
            nn++;
            
            if(total <= 0)
                break;
            
            while(nn != n)
            {
                v.push_back(matrix[nn][mm]);
                nn++;
                total --;
            }
            
            nn--;
            mm--;
            
           if(total <= 0)
                break;
            
            while(mm >= x)
            {
                v.push_back(matrix[nn][mm]);
                mm--;
                total --;
            }
            
            mm++;
            nn--;
            
           if(total <= 0)
                break;
            
            while(nn >= x+1)
            {
                v.push_back(matrix[nn][mm]);
                nn--;
                total--;
            }
            
            
            
            m--;
            n--;
            x++;
            
        }
        return v;
    }
};

File: SpiralMatrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rowSize = len(matrix)
        colSize = len(matrix[0])
        totalSize = rowSize * colSize
        spiral = []
        offset = 0

        # if there is a number in the matrix dont exit
        while totalSize > 0:

            row = offset
            col = offset

            # getting the upper row
            while col < colSize:
                spiral.append(matrix[row][col])
                col += 1
                totalSize -= 1

            if totalSize <= 0:
                break

            row += 1
            col -= 1
            # getting the right colomn
            while row < rowSize:
                spiral.append(matrix[row][col])
                row += 1
                totalSize -= 1

            if totalSize <= 0:
                break

            col -= 1
            row -= 1
            # getting the lower row
            while col >= offset:
                spiral.append(matrix[row][col])
                col -= 1
                totalSize -= 1


            if totalSize <= 0:
                break

            

            row -= 1
            col += 1
            # getting the left colomn
            while row >= (offset + 1):
                spiral.append(matrix[row][col])
                row -= 1
                totalSize -= 1


            colSize -= 1
            rowSize -= 1
            offset += 1

        return spiral


